draft-ietf-jsonpath-base.md

```diff
@@ -1361,9 +1362,47 @@ Provisional registration? (standards tree only):
 
 # Security Considerations {#Security}
 
-This section gives security considerations, as required by {{RFC3552}}.
-
-
+Security considerations for JSONPath can stem from
+
+* attack vectors on JSONPath implementations, and
+* the way JSONPath is used in security-relevant mechanisms.
+
+## Attack vectors on JSONPath Implementations
+
+Historically, JSONPath has often been implemented by feeding parts of
+the query to an underlying programming language engine, e.g.,
+JavaScript.
+This approach is well known to lead to injection attacks and would
+require perfect input validation to prevent these attacks (see
+{{Section 12 of -json}} for similar considerations for JSON itself).
+Instead, JSONPath implementations need to implement the entire syntax
+of the query without relying on the parsers of programming language
+engines.
+
+Attacks on availability may attempt to trigger unusually expensive
+runtime performance exhibited by certain implementations in certain
+cases.
+(See {{Section 10 of -cbor}} for issues in hash-table implementations,
+and {{Section 8 of -iregexp}} for performance issues in regular
+expression implementations.)
+Implementers need to be aware that good average performance is not
+sufficient as long as an attacker can choose to submit specially
+crafted JSONPath queries that trigger surprisingly high, possibly
+exponential, CPU usage.
+
+## Attacks on Security Mechanisms that Employ JSONPath
+
+Where JSONPath is used as a part of a security mechanism, attackers
+can attempt to evoke unexpected behavior, or take advantage of
+differences in behavior between implementations that implement some
+protection and the implementations that are being protected by these.
+
+This also applies to underlying technologies such as UTF-8 (see
+{{Section 10 of -utf8}}), the Unicode character set, and JSON.
+A characteristic of JSON that can lead to varying results is the
+fact that JSON objects are unordered; therefore, the order in which
+results of a JSONPath query reflect the presence of JSON object
+members can vary with implementations.
```

Comment:
Fair enough, but why is this a security consideration?

Reply:
In any check/use situation.
(A protective application [that checks data to protect another application] cannot rely on the protected application [that uses the data] to make exactly the same recognition decisions.)

Reply:
Does this presuppose that the protected application needs to be protected from objects with members that appear in a particular order?  I'm really having trouble imagining a situation where this makes a difference that could constitute a vulnerability.

Reply:
Let's say the protective application needs to avoid data that say {"rocket": "launch"} ever reaching the protected application.  Let's say the two differ in the way they handle duplicate keys.

Or let's say the protected handles the JSON data as it comes in, with a streaming parser, giving the attacker way more leverage to create check/use situations.

(Generally speaking, the problem with "proof by lack of imagination" in security considerations is that the set of attackers collectively have way more imagination than we do.)